Chess/desk.cpp

```diff
@@ -42,8 +45,123 @@ Desk::Desk(QObject *parent) :
     for (int i=0; i<8; i++) {
         buffer_[6][i] = new Pawn(this, fcolor);
     }
+}
+
+
+bool Desk::grabFigure(const Cell &cell)
+{
+    Figure *curr_fig = getFigure(cell);
+
+    if (curr_fig == NULL)
+        return false;
+
+    if (curr_fig->getColor() != color_to_move_)
+        return false;
+
+    have_active_figure_ = true;
+
+    command_ = new Command(this);
+    command_->set_b_info(cell);
+
+    return true;
+}
+
+bool Desk::putDownFigure(const Cell &cell)
+{
+    if (!have_active_figure_)
+        return false;
+
+    command_->set_e_info(cell);
+
+    if (command_->valid()) {
+        command_->exec();
+        executed_commands_.push_back(*command_);
+
+        delete command_;
+        command_ = NULL;
+
+        switchActivePlayer();
+        have_active_figure_ = false;
+
+        return true;
+
+    } else {
+        interruptCommand();
+        return false;
+    }
+}
+
+void Desk::switchActivePlayer()
+{
+    color_to_move_ = (color_to_move_== WHITE) ? BLACK : WHITE;
+}
+
+void Desk::rollback_from_list()
+{
+    if (command_ == &executed_commands_.front())
+        return;
+
+    if (command_ == NULL)
+        command_ = &executed_commands_.back();
+    else
+        command_--;
+
+
+    command_->rollback();
+}
+
+void Desk::make_move_from_list()
+{
+    if (command_ == NULL)
+        return;
+
+    command_->exec();
+
+    if (command_ == &executed_commands_.back())
+        command_ = NULL;
+    else
+        command_++;
+}
+
+bool Desk::interruptCommand()
+{
+    if (command_ != NULL) {
+        delete command_;
+        command_ = NULL;
+
+        have_active_figure_ = false;
+        return true;
+    } else {
+        return false;
+    }
+}
+
+const std::vector<Command> &Desk::getState()
+{
+    return executed_commands_;
+}
 
+void Desk::restoreState(const std::vector<Command> &commands)
+{
+    executed_commands_.clear();
+
+    for (std::vector<Command>::const_iterator it=commands.begin();
+                                      it!=commands.end();
+                                      it++)
+    {
+        command_ = new Command(this);
+        command_->set_b_info(it->get_b_info());
+        command_->set_e_info(it->get_e_info());
 
+        if (!command_->valid())
+            throw std::exception();

```

Comment:
don't use std::exception dirrectly, derive you own error from it.
It also good to add some string explanation what went wrong into exception.